python/foxglove-sdk/src/websocket_server.rs

```diff
@@ -0,0 +1,193 @@
+use crate::errors::PyFoxgloveError;
+use foxglove::{
+    websocket::{Client, ClientChannelView, ServerListener},
+    WebSocketServer, WebSocketServerBlockingHandle,
+};
+use pyo3::{
+    prelude::*,
+    types::{PyBytes, PyDict},
+};
+use std::sync::Arc;
+use std::time;
+
+/// A mechanism to register callbacks for handling client message events.
+#[pyclass(name = "ServerListener", module = "foxglove")]
+pub struct PyServerListener {
+    listener: Py<PyAny>,
+}
+
+#[pymethods]
+impl PyServerListener {
+    #[new]
+    fn new(listener: Py<PyAny>) -> Self {
+        PyServerListener { listener }
+    }
+}
+
+/// A client connected to a running websocket server.
+#[pyclass(name = "Client", module = "foxglove")]
+pub struct PyClient {
+    #[pyo3(get)]
+    id: u32,
+}
+
+/// Information about a client channel.
+#[pyclass(name = "ClientChannelView", module = "foxglove")]
+pub struct PyClientChannelView {
+    #[pyo3(get)]
+    id: u32,
+    #[pyo3(get)]
+    topic: String,
```

Comment:
This will create a Python string on every access, and involve at least two copies. One from rust str to owned string and a second on access to python string.

I don't think it will matter much to performance, since these are tiny, but we could save one copy by storing this a Python string and doing the conversion upfront.

Thoughts? We don't have to do this now, it's not a breaking change.

I'm more concerned about setting a good example for future code that we write than this particular case.

Reply:
👍 I changed this to PyString and now construct that directly from the &str.